cql3/statements/select_statement.cc

```diff
@@ -514,25 +514,26 @@ indexed_table_select_statement::do_execute_base_query(
         auto& [merger, ranges_to_vnodes, concurrency] = query_state;
-        return repeat([this, &ranges_to_vnodes, &merger, &proxy, &state, &options, &concurrency, cmd, timeout]() {
+        auto base_command = ::make_lw_shared<query::read_command>(*cmd);
+        auto old_paging_state = options.get_paging_state();
+        if (old_paging_state) {
+            auto base_pk = generate_base_key_from_index_pk<partition_key>(old_paging_state->get_partition_key(),
+                    old_paging_state->get_clustering_key(), *_schema, *_view_schema);
+            if (_schema->clustering_key_size() > 0) {
+                assert(old_paging_state->get_clustering_key().has_value());
+                auto base_ck = generate_base_key_from_index_pk<clustering_key>(old_paging_state->get_partition_key(),
+                        old_paging_state->get_clustering_key(), *_schema, *_view_schema);
+                base_command->slice.set_range(*_schema, base_pk,
+                        std::vector<query::clustering_range>{query::clustering_range::make_starting_with(range_bound<clustering_key>(base_ck, false))});
+            } else {
+                base_command->slice.set_range(*_schema, base_pk, std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
+            }
+        }
+        return repeat([this, &ranges_to_vnodes, &merger, &proxy, &state, &options, &concurrency, base_command, timeout]() {
             // Starting with 1 range, we check if the result was a short read, and if not,
             // we continue exponentially, asking for 2x more ranges than before
             dht::partition_range_vector prange = ranges_to_vnodes(concurrency);
-            auto command = ::make_lw_shared<query::read_command>(*cmd);
-            auto old_paging_state = options.get_paging_state();
-            if (old_paging_state && concurrency == 1) {
-                auto base_pk = generate_base_key_from_index_pk<partition_key>(old_paging_state->get_partition_key(),
-                        old_paging_state->get_clustering_key(), *_schema, *_view_schema);
-                if (_schema->clustering_key_size() > 0) {
-                    assert(old_paging_state->get_clustering_key().has_value());
-                    auto base_ck = generate_base_key_from_index_pk<clustering_key>(old_paging_state->get_partition_key(),
-                            old_paging_state->get_clustering_key(), *_schema, *_view_schema);
-                    command->slice.set_range(*_schema, base_pk,
-                            std::vector<query::clustering_range>{query::clustering_range::make_starting_with(range_bound<clustering_key>(base_ck, false))});
-                } else {
-                    command->slice.set_range(*_schema, base_pk, std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
-                }
-            }
+            auto command = ::make_lw_shared<query::read_command>(*base_command);
```

Comment:
Something looks wrong to me here - or maybe I just don't understand here.
In the old code, for the first iteration (concurrency==1), the code copied "cmd" and modified it based on the old paging state. But when concurrency > 1, the code also copied "cmd", but didn't modify it.
In the new code, all iterations copy the same base_command. What am I missing?
Also, do we have any test that can verify that this code actually works as intended?

Reply:
Oh snap, you're right... that's what I get for pushing patches too late in the evening. Only the first command should be trimmed, the rest should be a copy of the original command and this patch does not make sense. My original (correct) intention was that if we have an old paging state and our query spans multiple partitions, the first subquery may only need to fetch a small slice of the partition, depending of where we ended last time, but all the next queries should use the original, full range. I remember that there are dtests for that, I'll try to find out which ones.